### PR TITLE
[Orca] Add type hint with tf2 estimator 

### PR DESCRIPTION
**Pull request by sgwhat**

### CODE REVIEW

File: python/orca/src/bigdl/orca/learn/tf2/ray_estimator.py [hunk 1]
```diff
@@ -129,11 +141,23 @@ def __init__(self,
 
         self.num_workers = len(self.remote_workers)
 
-    def fit(self, data, epochs=1, batch_size=32, verbose=1,
-            callbacks=None, validation_data=None, class_weight=None,
-            steps_per_epoch=None, validation_steps=None, validation_freq=1,
-            data_config=None, feature_cols=None,
-            label_cols=None):
+    def fit(self,
+            data: Union["SparkXShards", "Dataset", "ray.data.Dataset", Callable],
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/learn/tf2/ray_estimator.py:145]: Union["SparkXShards", "bigdl.orca.data.tf.data.Dataset",..
Missing DataFrame

File: python/orca/src/bigdl/orca/learn/tf2/ray_estimator.py [hunk 2]
```diff
@@ -297,9 +326,19 @@ def zip_func(worker, this_partition_refs, that_partition_refs):
                                                                       zip_func)
         return worker_stats
 
-    def evaluate(self, data, batch_size=32, num_steps=None, verbose=1,
-                 sample_weight=None, callbacks=None, data_config=None,
-                 feature_cols=None, label_cols=None):
+    def evaluate(self,
+                 data: Union["SparkXShards",
+                             "Dataset",
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/learn/tf2/ray_estimator.py:331]: Full Name on Dataset
Missing DataFrame

File: python/orca/src/bigdl/orca/learn/tf2/ray_estimator.py [hunk 3]
```diff
@@ -445,9 +484,17 @@ def transform_func(worker, shards_ref):
         spark_xshards = pred_shards.to_spark_xshards()
         return spark_xshards
 
-    def predict(self, data, batch_size=32, verbose=1,
-                steps=None, callbacks=None, data_config=None,
-                feature_cols=None, min_partition_num=None):
+    def predict(self,
+                data: Union["SparkXShards", "DataFrame", "Dataset"],
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/learn/tf2/ray_estimator.py:488]: PandasDataFrame and SparkDataFrame?
And line501 should also be checked.

File: python/orca/src/bigdl/orca/learn/tf2/pyspark_estimator.py [hunk 4]
```diff
@@ -261,9 +277,16 @@ def transform_func(iter, init_param, param):
 
         return result[0]
 
-    def evaluate(self, data, batch_size=32, num_steps=None, verbose=1,
-                 sample_weight=None, callbacks=None, data_config=None,
-                 feature_cols=None, label_cols=None):
+    def evaluate(self,
+                 data: Union["SparkXShards", "DataFrame", Callable],
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/learn/tf2/pyspark_estimator.py:281]: SparkDataFrame

File: python/orca/src/bigdl/orca/learn/tf2/pyspark_estimator.py [hunk 5]
```diff
@@ -360,9 +383,14 @@ def transform_func(iter, init_param, param):
                 lambda iter: transform_func(iter, init_params, params)).collect()
         return res[0]
 
-    def predict(self, data, batch_size=32, verbose=1,
-                steps=None, callbacks=None, data_config=None,
-                feature_cols=None):
+    def predict(self,
+                data: Union["SparkXShards", "DataFrame"],
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/learn/tf2/pyspark_estimator.py:387]: SparkDataFrame

File: python/orca/src/bigdl/orca/learn/tf2/pyspark_estimator.py [hunk 6]
```diff
@@ -360,9 +383,14 @@ def transform_func(iter, init_param, param):
...
+                batch_size: Optional[int]=None,
+                verbose: int=1,
+                steps: Optional[int]=None,
+                callbacks: Optional[List["Callback"]]=None,
+                data_config: Optional[Dict[str, Any]]=None,
+                feature_cols: Optional[List[str]]=None) -> Union["SparkXShards", "DataFrame"]:
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/learn/tf2/pyspark_estimator.py:393]: spark

File: python/orca/src/bigdl/orca/learn/tf2/ray_estimator.py [hunk 7]
```diff
@@ -37,21 +37,32 @@
...
 from bigdl.orca.ray import OrcaRayContext
 
+from typing import TYPE_CHECKING, Any, Dict, List, Callable, Union, Optional
+if TYPE_CHECKING:
+    import numpy as np
+    import tensorflow as tf
+    from tensorflow.python.saved_model.save_options import SaveOptions
+    from tensorflow.python.keras.callbacks import Callback
+    from tensorflow.python.keras.engine.training import Model
+    from pyspark.sql import DataFrame
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/learn/tf2/ray_estimator.py:50]: SparkDataFrame

File: python/orca/src/bigdl/orca/learn/tf2/ray_estimator.py [hunk 8]
```diff
@@ -445,9 +484,17 @@ def transform_func(worker, shards_ref):
...
+                data: Union["SparkXShards", "DataFrame", "Dataset"],
+                batch_size: Optional[int]=None,
+                verbose: Union[str, int]=1,
+                steps: Optional[int]=None,
+                callbacks: Optional[List["Callback"]]=None,
+                data_config: Optional[Dict[str, Any]]=None,
+                feature_cols: Optional[List[str]]=None,
+                min_partition_num: Optional[int]=None) -> Union["SparkXShards",
+                                                                "DataFrame",
+                                                                "Dataset"]:
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/learn/tf2/ray_estimator.py:497]: Dataframe

**Review by leonardozcm — COMMENTED** [review]
Should check DataFrame supporting status.

File: python/orca/src/bigdl/orca/learn/tf2/ray_estimator.py [hunk 9]
```diff
@@ -252,17 +278,20 @@ def fit(self, data, epochs=1, batch_size=32, verbose=1,
...
                 for i in range(self.num_workers):
-                    params["data_creator"] = self.process_ray_dataset(shards[i],
-                                                                      label_cols, feature_cols,
-                                                                      data_config)
-                    params["validation_data_creator"] = self.process_ray_dataset(val_shards[i],
-                                                                                 label_cols,
-                                                                                 feature_cols,
-                                                                                 data_config)
+                    params["data_creator"] = self.process_ray_dataset(
+                        shards[i],
```
Comment (lalalapotter) [inline python/orca/src/bigdl/orca/learn/tf2/ray_estimator.py:286]: Do we need to add some space in front of the code to pass the code style check?

File: python/orca/src/bigdl/orca/learn/tf2/ray_estimator.py [hunk 10]
```diff
@@ -537,7 +584,7 @@ def get_model(self, sample_input=None):
         return self._get_model_from_state(state, sample_input=sample_input)
 
     @enable_multi_fs_save
-    def save_checkpoint(self, checkpoint):
+    def save_checkpoint(self, checkpoint: str):
```
Comment (lalalapotter) [inline python/orca/src/bigdl/orca/learn/tf2/ray_estimator.py:587]: return value?

**Review by lalalapotter — APPROVED** [review]
Others LGTM.

File: python/orca/src/bigdl/orca/learn/tf2/pyspark_estimator.py [hunk 11]
```diff
@@ -124,7 +124,7 @@ def fit(self,
             validation_freq: int=1,
             data_config: Optional[Dict]=None,
             feature_cols: Optional[List[str]]=None,
-            label_cols: Optional[List[str]]=None) -> List[Dict[str, float]]:
+            label_cols: Optional[List[str]]=None) -> Dict[str, float]:
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/learn/tf2/pyspark_estimator.py:127]: In line145, return value statement missing
Also in ray_est.fit

File: python/orca/src/bigdl/orca/learn/tf2/pyspark_estimator.py [hunk 12]
```diff
@@ -286,7 +286,7 @@ def evaluate(self,
                  callbacks: Optional[List["Callback"]]=None,
                  data_config: Optional[Dict]=None,
                  feature_cols: Optional[List[str]]=None,
-                 label_cols: Optional[List[str]]=None) -> List[Dict[str, float]]:
+                 label_cols: Optional[List[str]]=None) -> Dict[str, float]:
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/learn/tf2/pyspark_estimator.py:289]: A Dict would be just fine

File: python/orca/src/bigdl/orca/learn/tf2/ray_estimator.py [hunk 13]
```diff
@@ -163,7 +163,8 @@ def fit(self,
             validation_freq: int=1,
             data_config: Optional[Dict]=None,
             feature_cols: Optional[List[str]]=None,
-            label_cols: Optional[List[str]]=None) -> List[Dict[str, float]]:
+            label_cols: Optional[List[str]]=None) -> Union[List[Dict[str, float]],
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/learn/tf2/ray_estimator.py:166]: The values returned by ray_est.fit is different from pyspark_est.fit?

**Review by leonardozcm — APPROVED** [review]
Others LGTM

**Pull request merged**